### PR TITLE
allow multiple vacc skip moves

### PR DESCRIPTION
**Pull request by edknock**

### CODE REVIEW

File: R/lancelot.R [hunk 1]
```diff
@@ -1960,6 +1933,88 @@ lancelot_parameters_vaccination <- function(N_tot,
...
+    stop("Require 0 values in 'vacc_skip_progression_rate' for vaccine strata
+         that have 0 values in 'vacc_skip_to'")
+  }
+  if (any(vacc_skip_to != 0 & vacc_skip_to <= vacc_classes + 1)) {
+    stop("Require vacc_skip_to[j] = 0 or vacc_skip_to[j] > j + 1")
+  }
+  vacc_skip_moves <- which(vacc_skip_to > 0)
+  vacc_skip_to_moves <- vacc_skip_to[vacc_skip_moves]
+  if (length(unique(vacc_skip_to_moves)) != length(vacc_skip_to_moves)) {
+    stop("Cannot have more than one vaccine skip move to the same stratum")
```
Comment (annecori) [inline R/lancelot.R:1978]: I think this is what you meant by no duplication in the documentation; I think it'd be clearer to say no duplication in the "to" ie you can't have two skip moves from two different vaccine strata into the same one.

File: tests/testthat/test-lancelot-vaccination.R [hunk 2]
```diff
@@ -3005,68 +3005,86 @@ test_that("run sensible vaccination schedule with waning and boosters", {
...
-                                   vacc_skip_from = 1L, vacc_skip_to = 3L,
-                                   vacc_skip_weight = -1),
+                                   vacc_skip_to = c(3, 0, 0),
+                                   vacc_skip_weight = c(1.5, 0, 0)),
                "'vacc_skip_weight' must lie in")
+
+  expect_error(lancelot_parameters(1, "london",
+                                   rel_susceptibility = c(1, 1, 1, 1),
+                                   vacc_skip_to = c(3, 4, 0, 0)),
+               "Cannot have overlapping vaccine skip moves")
```
Comment (annecori) [inline tests/testthat/test-lancelot-vaccination.R:3051]: hmm so maybe I didn't understand the overlapping issue in the end. Would be good to clarify what this means

File: tests/testthat/test-lancelot-vaccination.R [hunk 3]
```diff
@@ -3005,68 +3005,86 @@ test_that("run sensible vaccination schedule with waning and boosters", {
...
-  expect_equal(p$vacc_skip_dose, 0)
+                           vacc_skip_to = c(0L, 4L, 0L, 0L, 0L),
+                           vacc_skip_weight = rep(0, 5),
+                           vacc_skip_progression_rate = c(0, 2, 0, 0, 0))
+  expect_equal(p$vacc_skip_dose, integer(3))
+  expect_equal(p$vacc_skip_dose_inverse, integer(5))
+  expect_equal(p$vacc_skip_dose_weight, rep(0, 3))
+  expect_equal(p$vacc_skip_from, c(0L, 0L, 0L, 2L, 0L))
+  expect_equal(p$vacc_skip_progression_rate_base, c(0, 2, 0, 0, 0))
+  expect_equal(p$vacc_skipped, c(0L, 2L, 2L, 0L, 0L))
```
Comment (annecori) [inline tests/testthat/test-lancelot-vaccination.R:3087]: I don't understand what is in vacc_skipped I think

File: tests/testthat/test-lancelot-vaccination.R [hunk 4]
```diff
@@ -3185,21 +3200,27 @@ test_that("Vaccine skip works in all compartments", {
   y <- mod$simulate(1:200 * 4)
   y <- mod$transform_variables(drop(y))
 
-  ## cum_n_X_vacc_skip should be equal to the totals in vaccine class 5
+  ## cum_n_X_vacc_skip[, 3, , ] should be equal to the totals in vaccine class 5
+  ## but there should be no other vacc skip moves
   expect_true(any(y$cum_n_S_vacc_skip > 0))
-  expect_equal(y$cum_n_S_vacc_skip,
+  expect_true(any(y$cum_n_S_vacc_skip[, c(1, 2, 4, 5), , ] == 0))
```
Comment (annecori) [inline tests/testthat/test-lancelot-vaccination.R:3206]: should this be all rather than any? 
Same in all the ones below

**Review by annecori — CHANGES_REQUESTED** [review]

File: R/lancelot.R [hunk 5]
```diff
@@ -260,7 +260,7 @@ NULL
 ##'
 ##' @param vaccine_index_dose2 The index to use for the second dose
```
[the remark this answers is not in the document]
Comment (edknock) [reply R/lancelot.R:261]: Agree it would be ideal ultimately

File: tests/testthat/test-lancelot-vaccination.R (hunk 4 above)
Comment (edknock) [reply tests/testthat/test-lancelot-vaccination.R]: Fixed

File: R/lancelot.R (hunk 1 above)
Comment (edknock) [reply R/lancelot.R:1984]: I've tried to make this clearer in the parameter description

File: R/lancelot.R [hunk 6]
```diff
@@ -270,28 +270,26 @@ NULL
...
-##'   are weighted in dose distribution relative to those in `vacc_skip_to - 1`.
-##'   Must be between 0 and 1. Must be 0 (which is the default) if
-##'   `vacc_skip_to == vacc_skip_from`
+##' @param vacc_skip_to A vector of integers of length equal to the number of
+##'   vaccine strata. The value of `vacc_skip_to[j]` represents the vaccine
+##'   stratum an individual in vaccine stratum `j` can move to in a
+##'   "vaccine skip move". Values can be 0 (which mean there is no vaccine skip
+##'   move from that stratum), or greater than or equal to `j + 2` (we already
+##'   account for moves between successive vaccine strata). Overlapping vaccine
+##'   skip moves are not allowed. Default is a vector of 0s
```
[the remark this answers is not in the document]
Comment (edknock) [reply R/lancelot.R]: I've tried to make this clearer and offer acceptable/unacceptable examples

**Review by annecori — APPROVED** [review]

**Pull request merged**